File: BEmu/ReferenceDataRequest/RequestReferenceElementOverride.cs
//------------------------------------------------------------------------------
// <copyright project="BEmu" file="ReferenceDataRequest/RequestReferenceElementOverride.cs" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------

namespace BEmu.ReferenceDataRequest
{
    using System;
    using System.Collections.Generic;
    using System.Linq;
    using System.Text;

    internal class RequestReferenceElementOverride : Element
    {
        private ElementReferenceString _fieldId;
        private ElementReferenceObject _value;

        internal RequestReferenceElementOverride()
        {
            this._fieldId = new ElementReferenceString("fieldId", "");
            this._value = new ElementReferenceObject("value", "");
        }

        public override void SetElement(string name, object value)
        {
            var toUse = name.ToUpper();
            switch (toUse)
            {
                case "FIELDID":
                    this._fieldId = new ElementReferenceString(toUse, value.ToString());
                    break;
                case "VALUE":
                    this._value = new ElementReferenceObject(toUse, value);
                    break;
                default:
                    throw new ArgumentException();
            }            
        }

        public override Name Name { get { return new Name("overrides"); } }
        public override int NumValues { get { return 1; } }
        public override int NumElements { get { return 2; } }
        public override object this[int index] { get { return null; } }

        internal override StringBuilder PrettyPrint(int tabIndent)
        {
            string tabs = Types.IndentType.Indent(tabIndent);
            StringBuilder result = new StringBuilder();

            result.AppendFormat("{0}{1} = {{{2}", tabs, this.Name, Environment.NewLine);
            result.Append(this._fieldId.PrettyPrint(tabIndent + 1));
            result.Append(this._value.PrettyPrint(tabIndent + 1));
            result.AppendFormat("{0}}}{1}", tabs, Environment.NewLine);

            return result;
        }

        public override IEnumerable<Element> Elements
        {
            get
            {
                yield return this._fieldId;
                yield return this._value;
            }
        }
        
        public override Element this[string name]
        {
            get
            {
                switch (name.ToUpper())
                {
                    case "FIELDID":
                        return this._fieldId;
                    case "VALUE":
                        return this._value;
                }
                return base[name];
            }
        }

        public override object this[string name, int index]
        {
            get
            {
                if (index == 0)
                    return this[name].GetValue();
                return base[name, index];
            }
        }

    }
}

File: BEmuJava/bemu/src/main/java/com/bemu/BEmu/ReferenceDataRequest/RequestReferenceElementOverride.java
package com.bemu.BEmu.ReferenceDataRequest;

import com.bemu.BEmu.Name;
import com.bemu.BEmu.Element;

public class RequestReferenceElementOverride extends ElementParent
{
    private ElementReferenceString _fieldId;
    private ElementReferenceObject _value;
    
    RequestReferenceElementOverride()
    {
        this._fieldId = new ElementReferenceString("fieldId", "");
        this._value = new ElementReferenceObject("value", "");
    }
    
    public void setElement(String name, Object value) throws Exception
    {
    	if(name.equals("fieldId"))
            this._fieldId = new ElementReferenceString(name, value.toString());
    	
    	else if(name.equals("value"))
    		this._value = new ElementReferenceObject(name, value);
    	
    	else
    		super.setElement(name, value);
    }
    
    public Element getElement(String name) throws Exception
    {
    	if(name.equals("fieldId"))
            return this._fieldId;
    	
    	else if(name.equals("value"))
    		return this._value;
    	
    	else
    		return super.getElement(name);
    }
    
    public Name name() throws Exception
    {
    	return new Name("overrides");
    }
    
    public int numValues()
    {
    	return 1;
    }
    
    public int numElements()
    {
    	return 2;
    }
    
    protected StringBuilder prettyPrint(int tabIndent) throws Exception
    {
        String tabs = com.bemu.BEmu.types.IndentType.Indent(tabIndent);
        StringBuilder result = new StringBuilder();

        result.append(String.format("%s%s = {%s", tabs, this.name().toString(), System.getProperty("line.separator")));
        result.append(this._fieldId.prettyPrint(tabIndent + 1));
        result.append(this._value.prettyPrint(tabIndent + 1));
        result.append(String.format("%s}%s", tabs, System.getProperty("line.separator")));
        
        return result;
    }

}
